### ISSUE TITLE
Backdrop CMS: Write ddev-specific settings in settings.local.php instead of directly to settings.php

### ISSUE THREAD

**Issue by klonos** [ddev/ddev#5747]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

As it is currently, ddev writes this bit at the very end of the default `settings.php` file in Backdrop:
```php
// Automatically generated include for settings managed by ddev.
$ddev_settings = dirname(__FILE__) . '/settings.ddev.php';
if (getenv('IS_DDEV_PROJECT') == 'true' && is_readable($ddev_settings)) {
  require $ddev_settings;
}
```

Right before this, the default `settings.php` file in Backdrop contains the following:
```php
/**
 * Include a local settings file, if available.
 *
 * To make local development easier, you can add a settings.local.php file that
 * contains settings specific to your local installation, or to any secondary
 * environment (staging, development, etc).
 *
 * Typically used to specify a different database connection information, to
 * disable caching, JavaScript/CSS compression, re-routing of outgoing emails,
 * Google Analytics, and other things that should not happen on development and
 * testing sites.
 *
 * This local settings file can be ignored in your Git repository, so that any
 * updates to settings.php can be pulled in without overwriting your local
 * changes.
 *
 * Keep this code block at the end of this file to take full effect.
 */
if (file_exists(__DIR__ . '/settings.local.php')) {
  include __DIR__ . '/settings.local.php';
}
```

As the comments explain, keeping your `settings.php` file "clean" makes it easier/safer to pull and push changes without having to worry about overwriting things.

### Describe your solution

Writing the ddev-specific settings in `settings.local.php` will have no effect on current functionality (`settings.ddev.php` will still be loaded - only instead of directly from `settings.php` it will be loaded from `settings.local.php`), while at the same time:
- both `settings.ddev.php` as well as `settings.local.php` will be git-ignored (since Backdrop git-ignores `/settings*.php` by default - see https://github.com/backdrop/backdrop/blob/1.x/.gitignore#L5)
- `settings.php` will be left alone, and people can do things like `git add .` without having to worry that they might be including changes that should not be committed.

**Comment by laryn** [comment]
Great idea.

**Comment by rfay** [comment]
The reason we use `settings.ddev.php` is that DDEV owns it. DDEV tries to make as few changes as possible to anything about your system or project. And `settings.local.php` is intended for changes custom to the system it's on (and not to be in source control). 

In the very earliest days of DDEV, we used `settings.local.php` for this, but changed to `settings.ddev.php` so that it was clear what DDEV was doing. 

We won't be altering settings.local.php, or putting settings in it. We have long experience with the current approach and it works great. Happy to continue the conversation.

**Comment by laryn** [comment]
I think the request to keep `settings.ddev.php` as is, but to write the following section into `settings.local.php` rather than `settings.php`:

```
// Automatically generated include for settings managed by ddev.
$ddev_settings = dirname(__FILE__) . '/settings.ddev.php';
if (getenv('IS_DDEV_PROJECT') == 'true' && is_readable($ddev_settings)) {
  require $ddev_settings;
}
```

**Comment by rfay** [comment]
Does settings.local.db always exist in Backdrop? Does it always get included (it does *not* in Drupal)

**Comment by laryn** [comment]
It doesn't always exist, but the check for it is in `settings.php` by default:

 - https://github.com/backdrop/backdrop/blob/1.x/settings.php#L504-L506

**Comment by rfay** [comment]
I've always wondered why Drupal has this commented out by default; Backdrop's approach is better. 

It would be a pretty easy change to backdrop.go I think, contribution welcome, as long as it doesn't break anything. It seems like it might do a double-include on existing projects.

**Comment by laryn** [comment]
Would it be too much technical debt to check if it's in `settings.php` (like it currently does) as well as checking `settings.local.php` before writing it to `settings.local.php`? Or is it not a huge problem if it's double included on existing projects?

**Comment by klonos** [comment]
> It would be a pretty easy change to backdrop.go I think, contribution welcome ...

@rfay I personally don't speak go, but I could try, plus there may be members in our community that do. To assist us with this, can you please at least point to the proper file where these changes would need to be made? (thanks in advance)

**Comment by klonos** [comment]
...OK, found it (I think): https://github.com/ddev/ddev/blob/master/pkg/ddevapp/backdrop.go

**Comment by klonos** [comment]
...another few questions:
- is the https://github.com/ddev/ddev/blob/master/pkg/ddevapp/drupal/backdrop/settings.php file being kept up to date with Backdrop core upstream? (or rather _meant_ to be kept up to date?)
- if we are to remove the default `// Automatically generated include for settings managed by ddev.` section and move that to `settings.local.php`, then is that file needed elsewhere?

**Comment by klonos** [comment]
...I've started a draft PR here: https://github.com/ddev/ddev/pull/5770 (I hope I've followed the title/description rules properly - if not, then please point out what I might have done wrong)

I'll need to understand more of how go works as well as the logic in backdrop.go in order to continue. If anyone else is more knowledgeable, then please feel free to pick up where I left.

**Comment by klonos** [comment]
For the moment, I've tweaked my bash scripts to "manually" move the ddev-specific settings from `settings.php` into `settings.local.php` and then also run `ddev config --disable-settings-management` to stop DDEV from re-adding them after each restart.

Thinking about it further, it would be nice to have an in-built setting plus a `--settings-management-file` parameter that you can use with `ddev config` 🤔 ...something like `ddev config --settings-management-file=settings.local.php` (the setting could default to `settings.php` when project type is `drupal` and `settings.local.php` if the project type is `backdrop`).

**Comment by rfay** [comment]
You know about `disable_settings_management` config.yaml setting right? Many people (especially experts) don't need/want DDEV's settings management, or want to do things a different way. DDEV's settings management is just there to get new folks going fast, but is not intended to solve all problems.

**Comment by mxr576** [comment]
> DDEV's settings management is just there to get new folks going fast, but is not intended to solve all problems.

That is true, but as I am also newbie to DDEV, I would expect that if there is a settings.local.php then DDEV would not try to modify settings.php, because that is usually tracked in VCS and therefore it constantly creates overrides. This would make the transition smoother from other LA/EMP stacks. 

#Drupal10

**Comment by rfay** [comment]
@mxr576 welcome to DDEV. Is your comment about Backdrop or about Drupal 10? 

If it's about Backdrop, a PR remains welcome.

If it's about Drupal 10, it's not possible (at least as far as I know), because in Drupal's default settings.php, the include of `settings.local.php` is not enabled by default.

**Comment by mxr576** [comment]
You are right, in Drupal core, settings.local.php is not included by default in default.settings.php, which is usually the source of settings.php. However, I have seen two patterns in the past decade:

a) some kind of automatism (bash script or something similar, like DDEV is) uncommented those lines when an environment was created to warrant the inclusion of settings.local.php
b) a prepared settings.php was committed to VCS (as it should) and that already contained the inclusion of settings.local.php and many other more, if that was necessary. For example, see what Amazee.io is doing in their Lagoon Drupal starterkit: https://github.com/lagoon-examples/drupal-base/tree/main/assets https://github.com/amazeeio/drupal-integrations/blob/main/composer.json#L19-L22

So based on this experience, it seems reasonable to assuming that if there is a settings.local.php already exists in the filesystem then it exists for a reason, so DDEV could try to extend that by default instead of settings.php -- and there could be a configuration option for modifying this OOTB decision.

**Comment by rfay** [comment]
As mentioned, DDEV once did use settings.local.php. Most people don't configure a settings.local.php for Drupal on local dev, and the current approach has worked quite well for many years.

Since Backdrop (the focus of this issue) does in fact load settings.local.php by default, I'm open to that approach, but it probably has its own drawbacks of dealing with an already-existing settings.local.php.

**Comment by mxr576** [comment]
> but it probably has its own drawbacks of dealing with an already-existing settings.local.php.

That is also a valid point. 


My current workaround:

```yml
# In config.local.yml
hooks:
  post-start:
    # Rollback auto-include added by ddev to settings.php, move it to
    # settings.local.php instead.
    - exec: "git checkout web/sites/default/settings.php"
    - exec: |
        if [ ! -f web/sites/default/settings.local.php ]; then echo '<?php
        // Include automatically generated configuration by ddev.
        $ddev_settings = __DIR__ . "/settings.ddev.php";
        if (getenv("IS_DDEV_PROJECT") == "true" && is_readable($ddev_settings)) {
          require $ddev_settings;
        }

        // Include project.local.settings.php with project-specific default local
        // overrides for development environments.
        $local_project_overrides = __DIR__ . "/project.local.settings.php";
        if (is_readable($local_project_overrides)) {
          include $local_project_overrides;
        }' > web/sites/default/settings.local.php; fi

```

**Comment by rfay** [comment]
All good. And never forget that DDEV doesn't insist on managing your settings in the first place. https://ddev.readthedocs.io/en/stable/users/usage/cms-settings/

[`ddev config --disable-settings-management`](https://ddev.readthedocs.io/en/stable/users/configuration/config/#type)

**Comment by mxr576** [comment]
Yes, but it generates a `settings.php` file with valuable pre-configurations, and if I’m not mistaken, that file always contains up-to-date tweaks. For example, when the Symfony Mailer configuration was added, I believe every project received it, even existing ones. Am I right?

Alternatively, I have considered opening a ticket to request a DDEV command that generates the `settings.ddev.php` file without performing any other actions.

**Comment by rfay** [comment]
With settings_management_disabled, DDEV does not generate a settings.php

**Comment by rfay** [comment]
Not sure how this ended up, but a PR is welcome as long as it doesn't mess with current Backdrop behavior, etc. Closing for now, happy to continue the conversation here or elsewhere.
